retrieval/scrape.go

```diff
@@ -446,6 +461,28 @@ func (sl *scrapeLoop) stop() {
 	<-sl.done
 }
 
+func (sl *scrapeLoop) processScrapeResult(samples model.Samples, scrapeErr error, start time.Time) error {
+	// Collect samples post-relabelling and label handling in a buffer.
+	buf := &bufferAppender{buffer: make(model.Samples, 0, len(samples))}
```

Comment:
Did you look at allocated space with heap profiling?
What where the GC stats? (In particular, how often did GC happen?)

In any case, I still don't get why we need to copy everything around one more time if it is so easy to avoid it for the case where no limitation is requested.